### DISCUSSION TITLE
3D mortar contact: why is contact stiffness matrix implemented by automatic differentiation?

### DISCUSSION THREAD

**Discussion by ch950423** [idaholab/moose#21220] [Q&A General] [answered]
Hi MOOSE experts.

I am trying to implement the 3D mortar contact using dual Lagrange multipliers following the [thesis](https://mediatum.ub.tum.de/1109994) of Alexander Popp. But the contact stiffness matrix is so complicated that it is even not fully written in the thesis. The thesis just told me every term of the contact stiffness matrix, i.e. the directional derivatives of the mortar matrix Δ**D** and Δ**M**.

I debugged the following example:
`moose/modules/contact/test/tests/3d-mortar-contact/frictionless-mortar-3d-test-derivative-trimming.i`

I found the contact stiffness matrix is computed as follows:
`moose/modules/contact/src/constraints/ADMortarLagrangeConstraint.C`
<pre>
90 void
91 ADMortarLagrangeConstraint::computeJacobian(Moose::MortarType mortar_type)
...
146  std::vector< DualReal > residuals_lower;
147  residuals_lower.resize(number_indices_on_lowerd, 0);
148
149  // Only populate nodal residuals on the primary/secondary surfaces
150  // We do this regardless of whether we are interpolating normals. Use of this class
151  // implies we have Lagrange elements, so internal (high-dimensional) normals have no meaning
152  // and should be zero. As such, we decide to omit them and avoid possible spurious population of
153  // automatic differentiation-generated derivatives.
154  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
155  {
156    unsigned int index_lower = 0;
157    for (const auto index : is_index_on_lower_dimension)
158    {
159      _i = index;
160      residuals_lower[index_lower] += _JxW_msm[_qp] * _coord[_qp] * computeQpResidual(mortar_type);
161
162      // Get rid of derivatives that we assume won't count (tolerance prescribed by user)
163      // residuals_lower[index_lower].derivatives().sparsity_trim(_ad_derivative_threshold);
164      index_lower++;
165    }
166  }
</pre>

The line 160 is computing the residuals and its derivatives, i.e. the contact forces and the contact stiffness. Then the `residuals_lower[i].derivatives()` are used for assembly of the contact stiffness matrix.

I found the `computeQpResidual(mortar_type)` is a `ADReal` and the automatic differentiation technique is used here.

Why is the contact stiffness matrix implemented by the automatic differentiation rather than implemented directly according to the formula? Is it because the formula is too complicated? What is the difference between the two ways?

What is your advise if I want to compute the contact stiffness matrix?

Thanks in advance.

**Comment by recuero** [comment, marked as answer]
The complexity of computing contact stiffness matrices is avoided by usage of automatic differentiation (AD) in MOOSE. AD also allows to change the way one defines contact forces without much development (e.g. one may want to introduce additional dependencies in the constraint equations). AD may generate additional computations to compute the system Jacobian, but, in the case of contact, since it acts on lower-dimensional domain, I have not found it to be limiting in any relevant way. I don't think there is a significant difference between the manual and the AD approach in the intended final result. Terms can be neglected both using AD or computing the Jacobian manually.

Not sure if you want to compute the contact stiffness matrix in your own code or get the result from MOOSE. In the former case, I'd suggest dig into the literature. In the latter, I'd output the Jacobian in MOOSE for simpler problems.

**Reply by ch950423** [reply]
I have read relevant literature [Automatic Differentiation in MetaPhysicL and Its Applications in MOOSE](https://www.tandfonline.com/doi/full/10.1080/00295450.2020.1838877) and [A mortar thermomechanical contact computational framework for nuclear fuel performance simulation](https://www.sciencedirect.com/science/article/abs/pii/S0029549322001625?via%3Dihub).

It seems the main feature of AD is the simplification of the computation of the derivative, at the cost of memory and computing efficiency, but without lose of accuracy. 

I still have some questions:

The relevant literature aforementioned focus on AD application in mortar contact in 2D.
What the effect of applying AD to 3D mortar frictional contact? How about the convergence?
When simulating a large model (i.e. too many DOFs and multiple contact zones existing), will the main disadvantages of AD show up? (will memory usage and computing efficiency become unacceptable?)

Thanks a lot.

**Reply by ch950423** [reply]
By the way, because the main complexity relies on computing the contact stiffness, is it feasible that I use AD to compute the contact stiffness matrix only, while computing other part of the system Jacobian manually (i.e. implementing other part of the system Jacobian directly according to the formula)?

**Reply by recuero** [reply]
> What the effect of applying AD to 3D mortar frictional contact? How about the convergence?

I don't think there is any fundamental difference between 2D and 3D, other than dimensionality, which would increase the number of AD dependencies by some factor.

> will the main disadvantages of AD show up? (will memory usage and computing efficiency become unacceptable?)

When the AD container size is required to be too large, some memory limitations could potentially come into play. Say, if you have many variables per element, 3D problem, high-order elements, and a fine (finely meshed relative to the secondary side) primary side which increases the number of dependencies via projections. Typical simulations do not reach this kind of problems.

> By the way, because the main complexity relies on computing the contact stiffness, is it feasible that I use AD to compute the contact stiffness matrix only, while computing other part of the system Jacobian manually (i.e. implementing other part of the system Jacobian directly according to the formula)?

Yes, that's what I meant by "AD may generate additional computations to compute the system Jacobian, but, in the case of contact, since it acts on lower-dimensional domain, I have not found it to be limiting in any relevant way".

**Reply by GiudGiud** [reply]
btw dont forget to reconfigure then recompile MOOSE with a larger ad-derivative-container size (see ./configure --help) to work on mortar contact with AD. 
Start with 200 maybe and let us know if you run into any issue

**Reply by ch950423** [reply]
OK
